python/modules/SoftBDeepAK8SFProducer.py

```diff
@@ -187,6 +184,43 @@ def GetDeepAK8SF(self, fjets):
         # print (top_sf, top_sferr, top_fastsf, top_fastsferr)
         return top_sf, top_sferr, top_fastsf, top_fastsferr
 
+    class TTreeReaderArrayWrapper:
+        def __init__(self, ttarray):
+            self.ttarray = ttarray
+
+        def __iter__(self):
+            for i in xrange(len(self.ttarray)):
+                yield self.ttarray[i]
+            return
+
+    def nGenParts(self, event):
+        if not self.isData:
+            GenPart_pdgId = np.fromiter(self.TTreeReaderArrayWrapper(event.GenPart_pdgId), int)
+            GenPart_statusFlags = np.fromiter(self.TTreeReaderArrayWrapper(event.GenPart_statusFlags), int)
+            GenPart_eta = np.fromiter(self.TTreeReaderArrayWrapper(event.GenPart_eta), float)
+            GenPart_phi = np.fromiter(self.TTreeReaderArrayWrapper(event.GenPart_phi), float)
+
+            fatJet_eta = np.fromiter(self.TTreeReaderArrayWrapper(event.FatJet_eta), float)
+            fatJet_phi = np.fromiter(self.TTreeReaderArrayWrapper(event.FatJet_phi), float)
+
+            # statusFlag 0x2100 corresponds to "isLastCopy and fromHardProcess"
+            # statusFlag 0x2080 corresponds to "IsLastCopy and isHardProcess"
+            genPartsFilt = (((abs(GenPart_pdgId) >= 1) & (abs(GenPart_pdgId) <= 5)) | (abs(GenPart_pdgId) == 21)) & (((GenPart_statusFlags & 0x2100) == 0x2100) | ((GenPart_statusFlags & 0x2080) == 0x2080))
```

Comment:
Here the nGenPart doesn't need to come from top quark? Can we store this variable per fatjet? Might be useful for other study.

Reply:
the mass doesn't care if the parton came from a top or not, so we make no such requirement here

Reply:
OK, do we want to store the nGenPart per fatjet?

Reply:
not a bad idea, I will push a change to add this